Model/FileIndex.php

```diff
@@ -0,0 +1,121 @@
+<?php
+
+namespace SomethingDigital\UpgradeHelper\Model;
+
+class FileIndex
+{
+    const THEME_OVERRIDE = 'theme_override';
+
+    const MODULE_OVERRIDE = 'module_override';
+
+    const INTERESTING_EXTENSIONS = [
+            'phtml',
+            'js',
+            'html',
+            'less'
+    ];
+
+    const WHITELISTED_BASENAMES = [
+        'requirejs-config.js'
+    ];
+
+    /**
+     * @var array[]
+     */
+    private $index;
+
+    public function __construct()
+    {
+        $this->index = [
+            self::THEME_OVERRIDE => [],
+            self::MODULE_OVERRIDE => [],
+        ];
+    }
+
+    public function populateIndex(): void
+    {
+        $this->populateIndexForPath(self::THEME_OVERRIDE, 'app/design/');
+        $this->populateIndexForPath(self::MODULE_OVERRIDE, 'vendor/');
+        $this->populateIndexForPath(self::MODULE_OVERRIDE, 'app/code/');
+    }
+
+    /**
+     * @param array $pathInfo
+     * @return array
+     */
+    public function getOverrideResults(array $pathInfo): array
+    {
+        $moduleResults = $this->getResultsByOverrideType($pathInfo, self::MODULE_OVERRIDE);
+        $themeResults = $this->getResultsByOverrideType($pathInfo, self::THEME_OVERRIDE);
+        return array_merge($moduleResults, $themeResults);
+    }
+
+    /**
+     * @param string $overrideType
+     * @param string $basePath
+     */
+    private function populateIndexForPath(
```

Comment:
I think we should add comment to this method. The `RecursiveDirectoryIterator` etc.. don't immediately click until you've really thought through what's going on. Just a high level comment explaining that this method builds out an index of file structure of the target Magento instance we're running against.